govuk-notify-spi/src/main/scala/uk/gov/nationalarchives/notifyspi/NotifyEmailSenderProvider.scala

```diff
@@ -0,0 +1,44 @@
+package uk.gov.nationalarchives.notifyspi
+
+import java.util
+
+import org.keycloak.email.EmailSenderProvider
+import org.keycloak.models.UserModel
+import uk.gov.service.notify.{NotificationClient, NotificationClientException}
+
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Success, Try}
+
+class NotifyEmailSenderProvider(notifyClient: NotificationClient) extends EmailSenderProvider {
+
+  private val templateIdKey = "GOVUK_NOTIFY_TEMPLATE_ID"
+
+  override def send(config: util.Map[String, String],
+                     user: UserModel,
+                     subject: String,
+                     textBody: String,
+                     htmlBody: String): Unit = {
+
+    val templateId: String = sys.env.get(templateIdKey) match {
+      case Some(id) => id
+      case _ => throw new NotificationClientException("Missing GovUk Notify templated id")
+    }
+
+    val personalisation: Map[String, String] = Map(
+      "keycloakMessage" -> textBody,
+      "keycloakSubject" -> subject)
+
+    Try {
+      notifyClient.sendEmail(
+        templateId,
+        user.getEmail,
+        personalisation.asJava,
+        user.getId)
+    } match {
+      case Failure(exception) => exception.printStackTrace
```

Comment:
Is this the way Keycloak expects you to handle failures? [EmailSenderProvider](https://www.keycloak.org/docs-api/7.0/javadocs/org/keycloak/email/EmailSenderProvider.html) has `throws EmailException` as part of its signature. Should you wrap the exception in [EmailException](https://www.keycloak.org/docs-api/7.0/javadocs/org/keycloak/email/EmailException.html) and throw that?

Reply:
Good point.